### DISCUSSION TITLE
Compilation using Visual Studio 2022

### DISCUSSION THREAD

**Discussion by sascha-holl** [vgvassilev/clad#970] [Q&A] [answered]
Did anybody successfully built clad on Windows? I'm failing to do so, since a couple of days now. I'm using LLVM 18.1.0, which I also built form source. When I do *not* add LLVM_ENABLE_PLUGINS and LLVM_EXPORT_SYMBOLS_FOR_PLUGINS for the LLVM compilation and compile clad with STATI, then compilation succeeds. However, I cannot execute any program without getting the console output "clad failed to place the generated derivative in the object. Make sure calls to clad are within a #pragma clad ON region."

So I thought the reason is that I need to "attach" the clad plugin to the clang compiler. But currently no DLL is generated. I only got a "cladPlugin.lib" and "cladDifferentiator.lib". So, I guess I need to add LLVM_ENABLE_PLUGINS and LLVM_EXPORT_SYMBOLS_FOR_PLUGINS. However, 2 projects fail then (due to DLL symbol export limit of 65535).

Is there any instruction how to use clad on Windows?

**Comment by vgvassilev** [comment]
Hi @de-sholl, Clang plugins are not supported on Windows, yet. @fsfod is actively working on this you can take a look at his recent work in llvm/llvm-project#96630. I hope in llvm 19 he succeeds to make LLVM_ENABLE_PLUGINS work for Windows.

That being said, Clad works on Windows in the [ROOT](https://root.cern) project. The trick is to build clang and statically link `cladPlugin.lib` and `cladDifferentiator.lib` to the clang binary.

**Reply by sascha-holl** [reply]
Hi @vgvassilev, thank you for your response. Sorry, don't get it yet. Never heard about the ROOT project. Why do I need it? Why can't I statically link `cladPlugin.lib` and `cladDifferentiator.lib` without it?

BTW: I want to use clad in C++, not Python. I'm especially interested in the smallpt-demo.

**Reply by vgvassilev** [reply]
Sorry for not being clear. You do not need ROOT. All you need is to statically link those libraries you got against the clang binary. This way clad will become part of clang statically and it will just work. Of course that will require modifying the clang build system.

**Reply by sascha-holl** [reply]
Thanks for clarification. LLVM is a huge project. I created the LLVM.sln as described [here](https://llvm.org/docs/GettingStarted.html#getting-the-source-code-and-building-llvm). How do I now statically link the two clad libraries in this solution? In which of the projects? And how will I use clad it in the application then? The invocations `-Xclang -add-plugin -Xclang clad ` are then not needed anymore, right?

**Reply by vgvassilev** [reply]
You will need `-DLLVM_ENABLE_PROJECTS="clang" -DLLVM_EXTERNAL_CLAD_SOURCE_DIR=path/to/clad` then somewhere in clang's or clad's CMakeLists you should add clad as a link dependency to clang.

> The invocations -Xclang -add-plugin -Xclang clad are then not needed anymore, right?

They are still needed.

**Reply by sascha-holl** [reply]
I already built the newest clad version using LLVM 18.1.0 and got `cladDifferentiator.lib` and `cladPlugin.lib` from that. Now you say I should rebuild LLVM. Okay, I've configured LLVM 18.1.0 now using `cmake -S llvm -B build -G "Visual Studio 17 2022 -DLLVM_ENABLE_PROJECTS="clang;lld" -DCMAKE_BUILD_TYPE=Release -DLLVM_EXTERNAL_CLAD_SOURCE_DIR=path/to/clad/include`. I'm not sure what `LLVM_EXTERNAL_CLAD_SOURCE_DIR` is expecting. I've provided the path to `clad/include` for now. Or should it be the root directory `clad/` instead?

However, it is not clear to me how I "add clad as a link dependency to clang". What I've tried is adding

`target_link_libraries(clang PRIVATE path/to/clad/out/build/x64-Clang-Release/lib/cladPlugin.lib  path/to/clad/out/build/x64-Clang-Release/lib/cladDifferentiator.lib)`

in `llvm-project-llvmorg-18.1.0/clang/tools/driver/CMakeLists.txt`, since there is the `clang_target_link_libraries` call for `clang`.

I'm not sure whether this is what you meant or not. Compilation succeeds for all projects except `ClangReplInterpreterTests.vcxproj` and `clang-repl.vcxproj`, but the reason is `fatal error LNK1189: library limit of 65535 objects exceeded` for both of them. Maybe that's not a problem for my intended use, but let me mention that without the changes I've described abvoe *all* projects compiled successfully before.

What's next? I've created a `clad_test.sln` console application, set the toolset to `LLVM` and provided a custom Directory.build.props pointing to my LLVM 18.1.0 folder. I've added `-Xclang -add-plugin -Xclang clad` to the command line:

![image](https://github.com/vgvassilev/clad/assets/33206178/f5fe6cca-db85-4b60-baaf-30c23a5e0fad)

But when I compile my project I get `unable to find plugin 'clad'`. Without the additional options, everything compiles, but I get the following console output when I try to compile your `BasicUsage.cpp`

> clad failed to place the generated derivative in the object
> Make sure calls to clad are within a #pragma clad ON region
> CladFunction is invalid
> Result is 0

Quite unfortunate that it is so hard to get this to work under Windows, since `clad` seems to be exactly what I need for an application of my next research project.

**Reply by vgvassilev** [reply]
>  Or should it be the root directory clad/ instead?

Yes, it should be the root directory.

> in llvm-project-llvmorg-18.1.0/clang/tools/driver/CMakeLists.txt, since there is the clang_target_link_libraries call for clang.

Yes, that's exactly what should be done.

> I'm not sure whether this is what you meant or not. Compilation succeeds for all projects except ClangReplInterpreterTests.vcxproj and clang-repl.vcxproj, but the reason is fatal error LNK1189: library limit of 65535 objects exceeded for both of them. Maybe that's not a problem for my intended use, but let me mention that without the changes I've described abvoe all projects compiled successfully before.

You do not need these. You can only build the target `clang` and you should be fine.

> But when I compile my project I get unable to find plugin 'clad'.

That's quite strange. Are you sure you have linked in clad's static libraries to clang?

> Quite unfortunate that it is so hard to get this to work under Windows, since clad seems to be exactly what I need for an application of my next research project.

You can always use the WSL which is essentially a linux subsystem on Windows, switch to a unix-based system or wait hopefully for llvm19 that has the mentioned changes to support plugins.

**Reply by sascha-holl** [reply]
> That's quite strange. Are you sure you have linked in clad's static libraries to clang?

I've added the `target_link_libraries` as I've shown you above. Is there anything else I was supposed to do?

**Reply by vgvassilev** [reply]
That should do what was supposed to. Do you see the libraries being listed on the link line in clang?

**Reply by sascha-holl** [reply]
Don't I somehow need to "register" what "clad" is? How should the compiler know only from statically linking the libraries? (I'm not familiar with LLVM/clang, so bear with me if that's a stupid question)

**Reply by sascha-holl** [reply]
What is "the link line in clang"? How can I figure this out?

**Reply by vgvassilev** [reply]
When building the build system shows some more information how the compiler was invoked.

**Reply by sascha-holl** [reply]
Wait a second, my exact output is

error: unable to find plugin 'clad'
1>C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.Cpp.ClangCl.Common.targets(255,5): error MSB6006: "clang-cl.exe" exited with code 1.

Doesn't that mean, the system is still invoking the clang build which shipped with Visual Studio?

I've set the platform toolset to LLVM:

![image](https://github.com/vgvassilev/clad/assets/33206178/8edeec01-42e3-45ea-8908-49f996cd71a8)

and the toolset version is also set correctly:

![image](https://github.com/vgvassilev/clad/assets/33206178/2a1e15bb-d869-4b6f-9117-5c06d6aeb640)

The `Directory.build.propos` is as I posted above. Is still something wrong here?

**Reply by vgvassilev** [reply]
I am more of a command line person rather than a VS person, not sure if I can help with that.

**Reply by sascha-holl** [reply]
I was able to check. This is the command line, which ins invoked in my application: `clang-cl.exe /c /Z7 /nologo /W3 /WX- /diagnostics:column /O2 /Oi /D NDEBUG /D _CONSOLE /D _UNICODE /D UNICODE /EHsc /MD /GS /Gy /fp:precise /std:c++17 /Fo"clad_test\x64\Release\\" /Gd /TP --target=amd64-pc-windows-msvc  -Xclang -add-plugin -Xclang clad main.cpp`.

**Reply by vgvassilev** [reply]
Okay, how about the compilation line for clang.exe? It needs to have those `cladPlugin.lib and `cladDifferentiator.lib` in the options.

**Reply by vgvassilev** [reply]
Another thing you could try out is our binder version of clad: https://notebooks.gesis.org/binder/jupyter/user/vgvassilev-clad-nhvd8vg3/doc/tree/demos/Jupyter/Intro.ipynb

**Reply by sascha-holl** [reply]
The IDE does not show me the actual invocation of the linker; even with the highest possible verbosity setting. However, I've manually added `cladPLugin.lib` and `cladDifferentiator.lib` to the linker inputs and they are shown in the clang project -> Linker -> Command Line -> Additional Options.

Is there any way to tell from the binary (clang.exe) if it worked or not? Maybe from a disassembler?

Your link above gives me a 403: forbidden btw.

**Reply by sascha-holl** [reply]
Why should the linker add (all of) `cladPlugin.lib` and `cladDifferentiate.lib` to the clang.exe at all? I think it will notice that there is no dependence between those libraries and the clang project and so they get excluded. It seems to be that something is missing.

**Reply by vgvassilev** [reply]
> The IDE does not show me the actual invocation of the linker; even with the highest possible verbosity setting. However, I've manually added `cladPLugin.lib` and `cladDifferentiator.lib` to the linker inputs and they are shown in the clang project -> Linker -> Command Line -> Additional Options.
> 
> Is there any way to tell from the binary (clang.exe) if it worked or not? Maybe from a disassembler?
> 
> Your link above gives me a 403: forbidden btw.

https://mybinder.org/v2/gh/vgvassilev/clad/master?labpath=%2Fdemos%2FJupyter%2FIntro.ipynb

**Reply by sascha-holl** [reply]
Thanks for the link. I still try to get things running. When I remove both `lib`s, the linker is claiming that he cannot find them. So, he's trying to link against them. My earlier guess was that the linker figures out that the libs are not used and hence excludes them. Now I've explicitly set "Link Library Dependencies" to true:

![image](https://github.com/vgvassilev/clad/assets/33206178/1db0ef11-61da-4206-be9b-ef77448cec24)

Unfortunately, I still get `unable to find plugin 'clad'`. (I also tried setting `Use Library Dependency Inputs` to `yes` as well; but that didn't work either).

 According to [this page](https://clang.llvm.org/docs/ClangPlugins.html) there seems to be an option to register the plugin dynamically. Should I give that approach a try?

I'd clearly prefer the static one, but I'm a bit lost about what I could try next. As I said before, I'm not familiar with clang, but isn't there a command line which lists all plugins clang is aware of? I mean, the error is it cannot find the plugin 'clad'. So it would be useful to know which plugins he knows about.

**Reply by vgvassilev** [reply]
This is the relevant commit I had in mind: https://github.com/root-project/root/pull/15030/commits/f2273e0b97bc223b374ab2e2d05414b23fff3494

We need to tell the linker to paste the whole static archive in the binary.

**Reply by sascha-holl** [reply]
What is your suggestion? Currently I'm working with the generated clang.sln. I thought "Link Library Dependencies" or "Use Library Dependency Inputs" might exactly do what you described; pasting the whole static archive into the binary.

Do you suggest to work with cmake instead? I've added the following to Clang -> CmakeLists.txt:

```
set(CLANG_PLUGIN_LINK_LIBS cladPlugin cladDifferentiator)
set(CLAD_LIBS "-WHOLEARCHIVE:cladPlugin.lib -WHOLEARCHIVE:cladDifferentiator.lib")
add_dependencies(clang clad)
target_link_libraries(clang PUBLIC ${CLANG_PLUGIN_LINK_LIBS})
```

**Reply by vgvassilev** [reply]
Yeah, maybe that would work. The sln file is generated by cmake. That should probably do.

**Reply by sascha-holl** [reply, marked as answer]
It finally works! I had to add `/WHOLEARCHIVE:cladPlugin.lib /WHOLEARCHIVE:cladDifferentiator.lib` to Linker -> Command Line -> Additional Options in the clang.sln. Now compilation and also execution of the BasicUsage examples works without an error.

Thank you very much for your help along!
